update retrieve_data in era5.py

TimFuermann opened the pull request:
- changed download_format to .zip
- added entcrypting of .zip file from download and merging of contained .nc files
- add url

<!--
SPDX-FileCopyrightText: Contributors to atlite <https://github.com/pypsa/atlite>

SPDX-License-Identifier: CC0-1.0
-->

Closes # (if applicable).

## Changes proposed in this Pull Request


## Checklist

- [ ] Code changes are sufficiently documented; i.e. new functions contain docstrings and further explanations may be given in `doc`.
- [ ] Unit tests for new features were added (if applicable).
- [ ] Newly introduced dependencies are added to `environment.yaml`, `environment_docs.yaml` and `setup.py` (if applicable).
- [ ] A note for the release notes `doc/release_notes.rst` of the upcoming release is included.
- [ ] I consent to the release of this PR's code under the MIT license.

TimFuermann committed:
update retrieve_data in era5.py

- changed download_format to .zip
- added entcrypting of .zip file from download and merging of contained .nc files
- add url

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

TimFuermann commented:
This pull request should solve the issue mentioned in #414

TimFuermann committed:
Update pyproject.toml

lkstrp commented on atlite/datasets/era5.py:
```diff
@@ -338,36 +340,76 @@ def retrieve_data(product, chunks=None, tmpdir=None, lock=None, **updates):
     """
-    request = {"product_type": "reanalysis", "format": "netcdf"}
+
+    # Set url for data download, this allows to switch to different data
+    # sources more easily.
+    url = "https://cds.climate.copernicus.eu/api"
```
How would you switch to other sources? Could you explain?
That should probably be a constant at module level, and maybe an optional argument

TimFuermann replied:
Hi Lukas, you are right, I just added it here for later improvement. For example, I have a local version in which I added the glofas data, but they are stored under a different url, it is not cds, but rather ads. Switching the url in the dataset is then just simpler then going to the "config" file of the cds-api. Furthermore, with a fixed url in the data retrieval, changes can be made more easily. It was just a comment here and will be taken up in future additions.

lkstrp replied:
Ok, that makes sense. But I don't think we want to hardcode the url in here without the ability to pass it as an argument. We are basically overriding any urls a user may have set up manually via the cds api config file without being able to change it. So just move it to an extra argument to get this merged. Future additions we can make anyways

lkstrp commented on atlite/datasets/era5.py:
```diff
@@ -338,36 +340,76 @@ def retrieve_data(product, chunks=None, tmpdir=None, lock=None, **updates):
 
     with lock:
-        fd, target = mkstemp(suffix=".nc", dir=tmpdir)
+        fd, target_zip = mkstemp(suffix=".zip", dir=tmpdir)
```
Do you have any experience on whether zipping always, even if `.nc' files could be downloaded, has any performance drawbacks?
I could imagine that it might even improve it, since the bottleneck is most likely download bandwidth, but I do not know.

TimFuermann replied:
My experience was that downloading .zip files directly, improves performance. Nevertheless, I did not test it.

lkstrp replied:
Which makes sense. That's good enough

lkstrp commented on pyproject.toml:
```diff
@@ -36,6 +36,7 @@ dependencies = [
     "numexpr",
     "xarray>=2024.03.0",
     "netcdf4",
+    "h5netcdf",
```
Why is `h5netcdf` added? Can it not also replace `netcdf4`? Not sure if we wanna have both as dependency and not one of them optional.

TimFuermann replied:
h5netcdf is somehow required, as the datastructure changed. It seems that the nc files within the zip folder use a different compression. There is so much going on at the CDS,, not completly clear from my side.
One comment on the forum mentions the general problem here: https://forum.ecmwf.int/t/problems-parsing-nc-files-after-outage/8060/3

For me netcdf4 was not working, I had to isntall h5netcdf, I could not directly find it at the ECMW forum, but it could be worth trying it with an update of the netcdf4 converter.

TimFuermann replied:
I found out that encoder=netcdf4, cannot directly be used for byte arrays, therefore h5netcdf was used.

lkstrp replied:
Ok. In my understanding `h5netcdf` can not be a full replacement for `netcdf4`. I'm fine with adding it as an dependency now. But leave a comment so we might wanna clean that up in future.

```suggestion
    "h5netcdf", # For new cds api. Maybe make optional / may not be needed with future netcdf4 updates
```

lkstrp reviewed (commented):
Thank you @TimFuermann ! I just have a couple of questions

TimFuermann committed:
Update retrieval times in era5 dataset

This needs to be done, as the new cds api takes only string values for a valid data call.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lkstrp commented:
Next to the comments, please add a release note. Afterwards we could merge this. Thank you !

TimFuermann committed:
Update pyproject.toml

Co-authored-by: Lukas Trippe <lkstrp@pm.me>

fneum commented on atlite/datasets/era5.py:
```diff
@@ -290,9 +292,9 @@ def retrieval_times(coords, static=False, monthly_requests=False):
-            "month": str(time[0].month),
-            "day": str(time[0].day),
+            "year": [str(time[0].year)],
+            "month": [str(time[0].month).zfill(2)],
```
If leading zeros are required, perhaps this is better (and "%d", "%Y" for day and year):
```suggestion
            "month": [time[0].strftime("%m")],
```

fneum reviewed (changes requested):
Thanks @TimFuermann!

As outlined in the corresponding issue, I would first like to reproduce the problem to see if this change is necessary.

Second thing is that I would like to see a brief performance check in terms of memory and time.

TimFuermann committed:
Apply suggestions from code review

Co-authored-by: Fabian Neumann <fabian.neumann@outlook.de>

TimFuermann committed:
Added strftime for better readability

Not tested so far.

maurerle commented:
As I read it, this PR can be closed in favor of the merged https://github.com/PyPSA/atlite/pull/454 :)

FabianHofmann commented:
closing as the issue is already resolved as mentioned above
